chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+  if (base::FeatureList::IsEnabled(features::kBraveEphemeralStorage)) {
+    storage = ephemeralSessionStorage(storage);
+  }
+
+  return storage;
+}
+
+StorageArea* BraveDOMWindowStorage::ephemeralSessionStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
+
+  LocalDOMWindow* window = GetSupplementable();
+  if (!window->IsCrossSiteSubframe())
+    return non_ephemeral_storage;
+
+  // If we were not able to create non-ephemeral session storage for this
+  // window, then don't attempt to create an ephemeral version.
+  if (!non_ephemeral_storage)
+    return nullptr;
+
+  if (ephemeral_session_storage_)
+    return ephemeral_session_storage_;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page, window);
+  if (!namespaces)
+    return nullptr;
+
+  auto storage_area =
+      namespaces->session_storage()->GetCachedArea(window->GetSecurityOrigin());
+
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_session_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
+                          StorageArea::StorageType::kSessionStorage);
+  return ephemeral_session_storage_;
+}
+
+StorageArea* BraveDOMWindowStorage::localStorage(
+    ExceptionState& exception_state) {
+  LocalDOMWindow* window = GetSupplementable();
+  auto* storage = DOMWindowStorage::From(*window).localStorage(exception_state);
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  if (base::FeatureList::IsEnabled(features::kBraveEphemeralStorage)) {
+    storage = ephemeralLocalStorage(storage);
+  }
+
   return storage;
 }
 
+StorageArea* BraveDOMWindowStorage::ephemeralLocalStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
+
+  LocalDOMWindow* window = GetSupplementable();
+  if (!window->IsCrossSiteSubframe())
+    return non_ephemeral_storage;
+
+  // If we were not able to create non-ephemeral local storage for this window,
+  // then don't attempt to create an ephemeral version.
+  if (!non_ephemeral_storage)
+    return nullptr;
+
+  if (ephemeral_local_storage_)
+    return ephemeral_local_storage_;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page, window);
+  if (!namespaces)
+    return nullptr;
+
+  auto storage_area =
+      namespaces->local_storage()->GetCachedArea(window->GetSecurityOrigin());
+
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_local_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
+                          StorageArea::StorageType::kSessionStorage);
```

Comment:
Some explanation of why it is session here would indeed be helpful for future readers

Reply:
:+1: